core/CronArchive.php

```diff
@@ -1851,12 +1912,18 @@ private function getUrlsWithSegment($idSite, $period, $date)
             $request = new Request($urlWithSegment);
             $logger = $this->logger;
             $self = $this;
-            $request->before(function () use ($logger, $segment, $segmentCount, &$processedSegmentCount, $idSite, $period, $urlWithSegment, $self) {
+            $request->before(function () use ($logger, $segment, $segmentCount, &$processedSegmentCount, $idSite, $period, $date, $urlWithSegment, $self) {
 
                 if ($self->isAlreadyArchivingSegment($urlWithSegment, $idSite, $period, $segment)) {
                     return Request::ABORT;
                 }
 
+                $isThereArchive = $this->isThereAValidArchiveForPeriod($idSite, $period, $date, $segment);
```

Comment:
@diosmosis probably not needed as part of this PR... do you reckon, we could dynamically change the lastN date that is applied in `$request->url` if something was invalidated? Like maybe `isThereAValidArchiveForPeriod ()` could return the `minInvalidatedArchiveDate`, based on this information we create the `lastN` string and then do something like `$request->url = str_replace(date=$date, date=newLastN, $request->url)`? Not sure it's clear what I mean? Just meaning in case we know we invalidated an older archive then we could adjust the date right away?

Reply:
Yes, I think it's possible, will look into it.

Reply:
Note: this code does the lastN queries on archive tables we were discussing on slack.

Reply:
@tsteur made the change, it will change the lastN to be `min(oldestInvalidatedDateDaysSinceToday, originalLastN)` (so it will still respect the max dates in CronArchive).